### ISSUE TITLE
Rule Proposal: `vue/no-unknown-elements`

### ISSUE THREAD

**Issue by mysticatea** [vuejs/eslint-plugin-vue#99]
## Please describe what the rule should do:

`no-unknown-elements` should report a use of unknown elements in templates.
Unknown elements are elements which satisfy all of the following conditions:

- It's not valid HTML elements.
- It's not valid SVG elements.
- It's not built-in components of Vue.js.
- The `components` property does not include it.
    - I guess this has limitations in static analytics :thinking: .
- It's not allowed by rule's option.
- Directive comments like `<!-- vue elements: foo, bar ,baz -->` does not include it.

This would be helpful to catch typo.

In fact, this `vue/unknown-elements` and `vue/no-undef` (#98) are the main purpose that I started to create this ESLint plugin.

### Options

```json
{
    "vue/no-unknown-elements": ["error", {
        "knownElements": ["foo", "bar", "baz"]
    }]
}
```

- `knownElements` (`string[]`) ... A list of known elements. Use for elements which are defined by Vue.js plugins.

## What category of rule is this? (place an "X" next to just one item)

[ ] Enforces code style
[X] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

## Provide 2-3 code examples that this rule will warn about:

```html
<template>
    <!-- div is well-known as a HTML element. -->
    <div></div>
    <!-- dav is unknown. -->
    <dav></dav>
</template>
```

```html
<template>
    <!-- foo is known since `components` property includes it. -->
    <foo></foo>
    <!-- fop is unknown. -->
    <fop></fop>
</template>
<script>
export default {
    components: {
        foo: {},
        bar: {}
    },
}
</script>
```

**Comment by armano2** [comment]
whats about `extends` in that case?

**Comment by armano2** [comment]
@mysticatea We should take into account `extends`:

```html
<template>
    <!-- foo is unknown but not reported due to `extends`. -->
    <foo></foo>
</template>
<script>
import AbstractComponent from './AbstractComponent.vue'

export default {
  extends: AbstractComponent
}
</script>
```

**Comment by mysticatea** [comment]
> We should take into account extends

Good point. I had forgotten it.

> also i will prefer to see html infix on all html rules to easily distinct them from each other.
eg. vue/no-unknown-html-elements

Hmm, if we use the name, I guess it's misleading that it disallows SVG or custom Vue components?
Do you know the case `<template lang=foo>` defines additional elements?

**Comment by armano2** [comment]
hmm, right we can use pug and other template systems :(

**Comment by michalsnik** [comment]
It makes sense, but then if you use for example `vuetify` or any other plugin that provides bunch of different components - adding them to config may be daunting and the rule will more likely be harmful.

**Comment by michalsnik** [comment]
This proposition seems like a candidate for many false positives. If we consider cases with `extend` and plugins that register some components on their own it would definitely cause more harm than good. That's why I'm closing this issue now. Let's focus on what we can actually do instead :)
